Explain Kubernetes auditing with Falco (#14332)
kubernetes/website: content/en/docs/tasks/debug-application-cluster/audit.md

let users route data where they want. For example, users can emit audit events to elasticsearch
plugin which supports full-text search and analytics.

## Webhook Collector Examples

### Use Falco to collect audit events

[Falco](falco_website) is an open source project for intrusion and abnormality detection for Cloud Native platforms.
This section describes how to set up Falco, how to send audit events to the Kubernetes Audit endpoint exposed by Falco, and how Falco applies a set of rules to automatically detect suspicious behavior.

> 

Reviewer: Remove excess whitespace.

#### Install Falco

Install Falco by using one of the following methods:

- [Standalone Falco](falco_installation)
- [Kubernetes DaemonSet](falco_installation)
- [Falco Helm Chart](falco_helm_chart)

> 

Reviewer: Remove excess whitespace.

Once Falco is installed make sure it is configured to expose the Audit webhook. To do so, use the following configuration:

> 

Reviewer: Remove excess whitespace.

let users route data where they want. For example, users can emit audit events to elasticsearch
plugin which supports full-text search and analytics.

## Webhook Collector Examples

### Use Falco to collect audit events

[Falco](falco_website) is an open source project for intrusion and abnormality detection for Cloud Native platforms.

> In this example, we will see how to send audit events to the Kubernetes Audit endpoint exposed by Falco.

Reviewer 2: ```suggestion
In this example, you will see how to send audit events to the Kubernetes Audit endpoint exposed by Falco.
```
(per [address the reader as ”you”](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you) from the documentation style guide).

After receiving Audit Events, Falco applies a set of rules, described below, in order to detect suspicious or notable behavior automatically.

#### Install Falco

You can install Falco standalone or as a Kubernetes DaemonSet following the instructions [here](falco_installation).
You can also install Falco using the official [Helm Chart](falco_helm_chart).

Once Falco is installed you need to make sure it is configured to expose the Audit webhook.

This is the configuration to enable it.

```yaml
webserver:
   enabled: true
   listen_port: 8765
   k8s_audit_endpoint: /k8s_audit
   ssl_enabled: false
   ssl_certificate: /etc/falco/falco.pem
```

This configuration is usually in `/etc/falco/falco.yaml`. When installed as a Kubernetes DaemonSet you can edit that file changing the `falco-config` ConfigMap.

#### Configure Kubernetes Audit

1. Create a [kubeconfig file](/docs/tasks/access-application-cluster/authenticate-across-clusters-kubeconfig/) for the [kube-apiserver][kube-apiserver] webhook audit backend

>         cat <<EOF > /etc/kubernetes/audit-webhook-kubeconfig

Author: Done, thanks!

        apiVersion: v1
        kind: Config
        clusters:
        - cluster:
            server: http://<ip_of_falco>:8765/k8s_audit
          name: falco
        contexts:
        - context:
            cluster: falco
            user: ""
          name: default-context
        current-context: default-context
        preferences: {}
        users: []
        EOF

2. Start [kube-apiserver][kube-apiserver] with the following options:

    ```shell
    --audit-policy-file=/etc/kubernetes/audit-policy.yaml --audit-webhook-config-file=/etc/kubernetes/audit-webhook-kubeconfig
    ```

#### Audit Rules

> Rules devoted to Kubernetes Audit Events can be found in [k8s_audit_rules.yaml][falco_k8s_audit_rules]. When installed as a native package or using the official Docker images, Falco installs this rules file to `/etc/falco/`, so they are available for use.

Author: Hey Tim, thanks a lot for your informal review! Really appreciated :)

Since the Falco rules file can change I don't think atm that copying it here reveals to be a good approach for readers.

- Any activity by a user outside of a set of allowed users, or by the anonymous user.
- Creating a pod with an image outside of a set of allowed images.
- Creating a privileged pod, a pod mounting a sensitive filesystem from the host, or a pod using host networking.
- Creating a NodePort service.
- Creating a configmap containing likely private credentials such as passwords, aws keys, etc.
- Attaching or execing to a running pod.
- Creating a namespace outside of a set of allowed namespaces.
- Creating a pod or service account in the kube-system or kube-public namespaces.
- Trying to modify or delete a system ClusterRole.
- Creating a ClusterRoleBinding to the cluster-admin role.
- Creating a ClusterRole with wildcarded verbs or resources (e.g. overly permissive)
- Creating a ClusterRole with write permissions or a ClusterRole that can exec to pods.

A second class of rules tracks resources being created or destroyed, including:

- Deployments
- Services
- ConfigMaps
- Namespaces
- Service accounts
- Role/ClusterRoles
- Role/ClusterRoleBindings

The final class of rules simply displays any Audit Event received by Falco. This rule is disabled by default, as it can be quite noisy.

> For further details refer to the this [Falco documentation page](falco_ka_docs).

Reviewer 2: Typically, Kubernetes documentation puts links like these inside a `whatsnext` capture block.

Author: We just used the same approach that logstash' links used Tim

let users route data where they want. For example, users can emit audit events to elasticsearch
plugin which supports full-text search and analytics.

## Webhook Collector Examples

### Use Falco to collect audit events

[Falco](falco_website) is an open source project for intrusion and abnormality detection for Cloud Native platforms.
In this example, you will see how to send audit events to the Kubernetes Audit endpoint exposed by Falco.

> After receiving Audit Events, Falco applies a set of rules, described below, in order to detect suspicious or notable behavior automatically.

Author: ```suggestion
```

#### Install Falco

You can install Falco standalone or as a Kubernetes DaemonSet following the instructions [here](falco_installation).

> You can also install Falco using the official [Helm Chart](falco_helm_chart).

Author: ```suggestion
```

Once Falco is installed you need to make sure it is configured to expose the Audit webhook.

> This is the configuration to enable it.

Author: ```suggestion
```

    - cluster:
        server: http://<ip_of_falco>:8765/k8s_audit
      name: falco
    contexts:
    - context:
        cluster: falco
        user: ""
      name: default-context
    current-context: default-context
    preferences: {}
    users: []
    EOF
    ```

2. Start [kube-apiserver][kube-apiserver] with the following options:

    ```shell
    --audit-policy-file=/etc/kubernetes/audit-policy.yaml --audit-webhook-config-file=/etc/kubernetes/audit-webhook-kubeconfig
    ```

#### Audit Rules

Rules devoted to Kubernetes Audit Events can be found in [k8s_audit_rules.yaml][falco_k8s_audit_rules]. When installed as a native package or using the official Docker images, Falco installs this rules file to `/etc/falco/`, so they are available for use.

There are three classes of rules.

The first class of rules looks for suspicious or exceptional activity. This includes things like:

- Any activity by a user outside of a set of allowed users, or by the anonymous user.
- Creating a pod with an image outside of a set of allowed images.
- Creating a privileged pod, a pod mounting a sensitive filesystem from the host, or a pod using host networking.
- Creating a NodePort service.

> - Creating a ConfigMap containing likely private credentials such as passwords, cloud provider secrets, etc.

Reviewer 3: ```suggestion
- Creating a ConfigMap containing private credentials, such as passwords and cloud provider secrets.
```

> - Attaching or execing to a running pod.

Reviewer 3: ```suggestion
- Attaching to or executing a command on a running pod.
```

> - Creating a namespace outside of a set of allowed namespaces.

Reviewer 3: ```suggestion
- Creating a namespace external to a set of allowed namespaces.
```

- Creating a pod or service account in the kube-system or kube-public namespaces.
- Trying to modify or delete a system ClusterRole.
- Creating a ClusterRoleBinding to the cluster-admin role.

> - Creating a ClusterRole with wildcarded verbs or resources (e.g. overly permissive)

Reviewer 3: ```suggestion
- Creating a ClusterRole with wildcarded verbs or resources. For example,  overly permissive.
```

> - Creating a ClusterRole with write permissions or a ClusterRole that can exec to pods.

Reviewer 3: ```suggestion
- Creating a ClusterRole with write permissions or a ClusterRole that can execute commands on pods.
```
